### ISSUE TITLE
Issues with delayed queue jobs processing

### ISSUE THREAD

**Issue by harshit0410** [OptimalBits/bull#2235]
#### Background
We have a rate limit configured on a bull queue (100 jobs per minute). We had a case where the number of incoming jobs exceeded the rate limit (maybe 200 jobs per minute) and the jobs started getting accumulated in the delayed queue as expected. After this event, the number of incoming jobs returned to normal (20 jobs per minute) which is far less than the rate limit we had configured.

#### Problems
We started seeing a delay in the processing of new incoming jobs and we also noticed that the delayed queue always had some delayed jobs in it even though we had limits configured far higher than that of incoming jobs + jobs in the delayed queue. This means that all the incoming jobs are getting processed with some delay which shouldn't be the case as limits are far higher.

On further investigation, we found a case in the bull codebase where the incoming jobs get delayed by some amount of time if there are some jobs in the delayed queue. Due to this, it becomes a cycle where incoming jobs are being delayed due to the existing jobs in the delayed queue.

#### Steps to reproduce

- Configure a bull queue with rate limiter config (150 jobs per minute).
- Run two scripts that add jobs to the queue at the rate of(60 jobs per minute, 120 jobs per minute).
- Observe the delayed queue getting filled as expected.
- Kill the script that adds jobs at (60 jobs per minute).
- Observe the size of the delayed queue reducing first and then getting stabilized roughly at the same rate as incoming jobs.
- The delayed queue never gets empty due to which incoming jobs are delayed as well no matter how much time we give the queue to process delayed jobs. 
- Increasing the bull queue rate limit config(max jobs) does not help.

## Bull version
V3.29.2

**Comment by manast** [comment]
Thanks for the reporting, I will try to reproduce this case following your instructions.

**Comment by grakhejapstmn** [comment]
is there any update on this? its still happening for us.

**Comment by manast** [comment]
I could reproduce it but I haven't had time to dig more into it.

**Comment by grakhejapstmn** [comment]
I believe the root cause lies somewhere in this script, 
https://github.com/OptimalBits/bull/blob/develop/lib/commands/moveToActive-8.lua#L64

We are using the size of the limited queue to determine how much delay to add to the incoming jobs. 
So if there's a sudden spike and we hit the rate limit once, since there are jobs in the waiting queue that will be processed after a given delay and the current rate is less than the limit it leads to a loop where all incoming jobs go the waiting queue, and since more jobs come in before the queue is empty all new jobs again go the waiting queue and the cycle repeats. 

I'm not sure what the fix could be here though. If we don't care about the order of jobs we can move jobs to the waiting list only if the rate limit is hit and otherwise process it. 
If we want to preserve the order maybe instead of depending on a delay to serially process the jobs in the waiting queue we can treat it like a priority queue and just pop items from it. 
[Sidekiq](https://github.com/mperham/sidekiq/wiki/Error-Handling#automatic-job-retry) adds a delay to the retried jobs only if they fail and bases the delay based on the number of retries which I feel is a lot more useful. Otherwise it processes them in the.

**Comment by manast** [comment]
> [Sidekiq](https://github.com/mperham/sidekiq/wiki/Error-Handling#automatic-job-retry) adds a delay to the retried jobs only if they fail and bases the delay based on the number of retries which I feel is a lot more useful. Otherwise it processes them in the order they came in.

This works exactly the same in Bull, that's the Backoff option and retry strategies. But the poster issue is not about delayed jobs per se, but on rate limited jobs, your case seems to be different. The best is if you create a separate issue with the specifics of your case.

**Comment by grakhejapstmn** [comment]
I realised I gave the incorrect example. But the source code I'm pointing to it the root cause. I'm from the same team that raised this issue and our issue is in fact related to rate limiting that's happening. 
Sidekiq has different strategies to solve for the concurrency problem (https://github.com/mperham/sidekiq/wiki/Ent-Rate-Limiting#bucket) but I'm not able to relate that to our issue at hand. 
Can you look into the lua script and see if there's something that can be done to fix the infinite loop without changing the architecture perhaps?

**Comment by manast** [comment]
@grakhejapstmn the best is if you can provide a piece of code in a new issue where you highlight the exact issue you are having, currently I am not able to understand your case. I prefer code as that is the most precise way to illustrate any issue.

**Comment by grakhejapstmn** [comment]
I feel the bug could be seen in this code in bull
```
 if isLimited == 1 then
     -- Remove from limited zset since we are going to try to process it
     rcall("SREM", limitedSetKey, jobId)
     rcall("HDEL", limiterIndexTable, jobId)
  else
    -- If not, check if there are any limited jobs
    -- If the job has not been rate limited, we should check if there are any other rate limited jobs, because if that
    -- is the case we do not want to process this job, just calculate a delay for it and put it to "sleep".
    local numLimitedJobs = rcall("SCARD", limitedSetKey)

    if numLimitedJobs > 0 then
      -- Note, add some slack to compensate for drift.
      delay = ((numLimitedJobs * ARGV[7] * 1.1) /  maxJobs) + tonumber(rcall("PTTL", rateLimiterKey))
    end
  end

  local jobCounter = tonumber(rcall("GET", rateLimiterKey))
  if(jobCounter == nil) then
    jobCounter = 0
  end
  -- check if rate limit hit
  if (delay == 0) and (jobCounter >= maxJobs) then
    -- Seems like there are no current rated limited jobs, but the jobCounter has exceeded the number of jobs for this unit of time so we need to rate limit this job.
    local exceedingJobs = jobCounter - maxJobs
    delay = tonumber(rcall("PTTL", rateLimiterKey)) + ((exceedingJobs) * ARGV[7]) / maxJobs
  end
```

If a job is already in the limited set the incoming jobs of the same type are directly pushed to waiting queue with a delay and only after. 

We know that this happens because our job throughput it much lesser than the limit that we have configured and now the waiting queue keeps getting filled and never goes to fully 0(As we have new jobs coming round the clock). We believe we crossed the limit once in December but this loop has been running since then and we've never been able to clear the waiting queue. This leads to a notable delay between when the action happened and when the job was actually performed.

Bull Monitor 
![image](https://user-images.githubusercontent.com/97720875/152727229-1d115c33-e6af-419e-b36d-58bf6341b80d.png)
Job Processing Throughput 
![image](https://user-images.githubusercontent.com/97720875/152727556-8e517d53-402b-4b18-ac4c-ce25cc9a354d.png)


Our limiter configuration: 
```
 limiter: {
      max: 15000,
      duration: 60000
    }
````

**Comment by casey-chow** [comment]
Is there a workaround for this issue besides re-architecting our processors? It seems important to call out this specific issue with a workaround because it makes the rate limiting functionality difficult to use as is, unless there's a way to fix the job queueing.

**Comment by manast** [comment]
@casey-chow unfortunately due to the current design in Bull it is very difficult to improve the rate-limiting without removing the "limit by group key" functionality. 

On the BullMQ Pro version(https://docs.bullmq.io/bullmq-pro/groups/rate-limiting), we improved the rate limit for groups in such a way that jobs are not moved around when a group is rate limited, instead, the workers are idling until the new job can be processed according to the rate limit.  It is possible then to simulate a global rate limit by only having 1 group for all jobs.
